cleverhans/attacks.py

```diff
@@ -1,62 +1,233 @@
-                       gamma, increase, clip_min, clip_max)
-    elif back == 'th':
-        raise NotImplementedError("Theano jsma not implemented.")
+class Attack:
+    """
+    Abstract base class for all attack classes.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, x, pred, y=None, backend='tf', clip_min=None, clip_max=None):
+        """
+        :param x: A placeholder for the model inputs.
+        :param pred: The model's symbolic output.
+        :param y: A placeholder for the model inputs. Only provide this parameter if
+                you'd like to use the true labels when crafting adversarial samples.
+                Otherwise, model predictions are used as labels to avoid the label
+                leaking effect. Default is None.
+        :param backend: A string indicating which backend to use. Must be either
+                        'tf' or 'th'. Default is 'tf'.
+        :param clip_min: Optional float parameter that can be used to set a minimum
+                        value for components of the example returned.
+        :param clip_max: Optional float parameter that can be used to set a maximum
+                        value for components of the example returned.
+        """
+        assert backend == 'tf' or backend == 'th'
+        self.x = x
+        self.pred = pred
+        self.y = y
+        self.backend = backend
+        self.clip_min = clip_min
+        self.clip_max = clip_max
+
+    def generate_symbolic(self):
+        """
+        Generate symbolic graph for adversarial samples and return. This
+        method should be overwritten in any child class which has a
+        symbolic implementation available.
+        """
+        raise NotImplementedError("The symbolic version of this attack "
+                                  "is not currently implemented.")
+
+    @abstractmethod
+    def generate_numpy(self, sess):
+        """
+        Generate adversarial samples and return them in a Numpy array.
+        """
+        if self.backend == 'tf':
+            # Using TensorFlow backend; must pass a session argument.
+            assert sess is not None
+        else:
+            # Using Theano backend; no session required.
+            assert sess is None
+
+
+class FastGradientMethod(Attack):
+    """
+    The Fast Gradient Method. This attack was originally implemented by
+    Goodfellow et al. (2015) with the infinity norm ("Fast Gradient
+    Sign Method").
+    Paper link: https://arxiv.org/pdf/1412.6572.pdf
+    """
+    def __init__(self, x, pred, y=None, backend='tf', clip_min=None, clip_max=None,
+                 eps=0.3, ord='inf'):
+        """
+        Create a FastGradientMethod instance.
+        :param eps: A float indicating the step size to use for the adversarial algorithm
+                    (input variation parameter).
+        :param ord: A string indicating the norm order to use when computing gradients.
```

Comment:
We may prefer here to use the same notation than numpy (np.inf, 1, 2) so we don't have to learn another convention